src/util/psbt/mod.rs

```diff
@@ -72,6 +72,24 @@ pub struct PartiallySignedTransaction {
 }
 
 impl PartiallySignedTransaction {
+    /// Returns an iterator for the funding UTXOs of the psbt
+    /// function throws error if transaction inputs have unequal lengths
+    pub fn get_funding_utxos(&self) -> impl Iterator<Item = Result<&TxOut, Error>> {
+        assert_eq!(self.inputs.len(), self.unsigned_tx.input.len());
+        self.unsigned_tx.input.iter().zip(&self.inputs).map(|(tx_input, psbt_input)| {
+            if let Some(ref witness_utxo) = psbt_input.witness_utxo {
+                    Ok(witness_utxo)
+                    // if no witness utxos then catch the non witness part of the input
+                } else if let Some(ref non_witness_utxo) = psbt_input.non_witness_utxo {
+                        // get output index (vout) from the outpoint and use it to get the non_witness_utxo
+                    let vout = tx_input.previous_output.vout as usize;
+                    Ok(&non_witness_utxo.output[vout])
```

Comment:
Blocker:

This code is vulnerable to out-of-bounds access if PSBT comes from untrusted source (e.g. BIP78 would be the case).
There's an easy fix:

```rust
non_witness_utxo.output.get(vout).ok_or(Error::PsbtUtxoOutOfbounds)
```

Obviously this requires adding another error variant.

Why it's OK to panic above but not here: the deserializer guarantees inputs will have equal length so that's generally safe and we even plan to enforce it statically. This is not the case here since the deserializer does **not** guarantee `vout` being in bounds.

Reply:
Wouldn't this cause arms mismatch ? 
at one arm `&txout` type, but in the second arm `Result<&txout, error>` type....